OpenRA.Mods.Common/Activities/Resupply.cs

```diff
@@ -56,46 +61,77 @@ public Resupply(Actor self, Actor host, WDist closeEnough)
 				activeResupplyTypes |= ResupplyType.Rearm;
 		}
 
-		protected override void OnFirstRun(Actor self)
-		{
-			if (host.Type == TargetType.Invalid)
-				return;
-
-			if (activeResupplyTypes > 0)
-				foreach (var notifyResupply in notifyResupplies)
-					notifyResupply.BeforeResupply(host.Actor, self, activeResupplyTypes);
-
-			// Reset the ReloadDelay to avoid any issues with early cancellation
-			// from previous reload attempts (explicit order, host building died, etc).
-			// HACK: this really shouldn't be managed from here
-			if (activeResupplyTypes.HasFlag(ResupplyType.Rearm))
-				foreach (var pool in rearmable.RearmableAmmoPools)
-					pool.RemainingTicks = pool.Info.ReloadDelay;
-		}
-
 		public override bool Tick(Actor self)
 		{
 			// HACK: If the activity is cancelled while we're already resupplying (or about to start resupplying),
-			// move actor outside the resupplier footprint
-			// TODO: This check is nowhere near robust enough, and should be rewritten
+			// move actor outside the resupplier footprint.
+			// TODO: This check is nowhere near robust enough, and should be rewritten.
 			if (IsCanceling && host.IsInRange(self.CenterPosition, closeEnough))
 			{
-				QueueChild(self.Trait<IMove>().MoveToTarget(self, host));
 				foreach (var notifyResupply in notifyResupplies)
 					notifyResupply.ResupplyTick(host.Actor, self, ResupplyType.None);
 
-				return false;
-			}
+				var aircraft = self.TraitOrDefault<Aircraft>();
+				if (aircraft != null)
+				{
+					aircraft.AllowYieldingReservation();
+					if (aircraft.Info.TakeOffOnResupply)
+						Queue(new TakeOff(self));
 
-			if (IsCanceling || host.Type == TargetType.Invalid
-				|| (closeEnough.LengthSquared > 0 && !host.IsInRange(self.CenterPosition, closeEnough)))
+					return true;
+				}
+				else if (self.Info.HasTraitInfo<MobileInfo>())
```

Comment:
Can save some trait(info) lookups here by doing
```
var move = self.Trait<IMove>();
var aircraft = move as Aircraft;
```

above, and then using `move` instead of `MobileInfo` and the lookups below.